File: ex-14/ex-14_java/Vetor2D.java
public class Vetor2D {
    private double x;
    private double y;

    public Vetor2D() {
        this.x = 0.0;
        this.y = 0.0;
    }

    public Vetor2D(double x, double y) {
        this.x = x;
        this.y = y;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public void setX(double x) {
        this.x = x;
    }

    public void setY(double y) {
        this.y = y;
    }

    public double magnitude() {
        return Math.sqrt(x * x + y * y);
    }

    public Vetor2D normalizado() {
        double mag = magnitude();
        if (mag > 0) {
            return new Vetor2D(x / mag, y / mag);
        }
        return new Vetor2D(0, 0);
    }

    public Vetor2D somar(Vetor2D outro) {
        return new Vetor2D(x + outro.x, y + outro.y);
    }

    public Vetor2D subtrair(Vetor2D outro) {
        return new Vetor2D(x - outro.x, y - outro.y);
    }

    public Vetor2D multiplicar(double escalar) {
        return new Vetor2D(x * escalar, y * escalar);
    }

    public Vetor2D dividir(double escalar) {
        if (escalar != 0) {
            return new Vetor2D(x / escalar, y / escalar);
        }
        return new Vetor2D(0, 0);
    }

    public Vetor2D somarIgual(Vetor2D outro) {
        x += outro.x;
        y += outro.y;
        return this;
    }

    public Vetor2D subtrairIgual(Vetor2D outro) {
        x -= outro.x;
        y -= outro.y;
        return this;
    }
}

File: ex-14/ex-14_python/vetor2d.py
import math


class Vetor2D:
    def __init__(self, x=0.0, y=0.0):
        self.x = x
        self.y = y

    def getX(self):
        return self.x

    def getY(self):
        return self.y

    def setX(self, x):
        self.x = x

    def setY(self, y):
        self.y = y

    def magnitude(self):
        return math.sqrt(self.x * self.x + self.y * self.y)

    def normalizado(self):
        mag = self.magnitude()
        if mag > 0:
            return Vetor2D(self.x / mag, self.y / mag)
        return Vetor2D(0, 0)

    def __add__(self, outro):
        return Vetor2D(self.x + outro.x, self.y + outro.y)

    def __sub__(self, outro):
        return Vetor2D(self.x - outro.x, self.y - outro.y)

    def __mul__(self, escalar):
        return Vetor2D(self.x * escalar, self.y * escalar)

    def __truediv__(self, escalar):
        if escalar != 0:
            return Vetor2D(self.x / escalar, self.y / escalar)
        return Vetor2D(0, 0)

    def __iadd__(self, outro):
        self.x += outro.x
        self.y += outro.y
        return self

    def __isub__(self, outro):
        self.x -= outro.x
        self.y -= outro.y
        return self
